include/dlpack/dlpack.h

```diff
@@ -216,13 +243,66 @@ typedef struct DLManagedTensor {
    *   which DLManagedTensor is used in the framework. It can also be NULL.
    */
   void * manager_ctx;
-  /*! \brief Destructor signature void (*)(void*) - this should be called
-   *   to destruct manager_ctx which holds the DLManagedTensor. It can be NULL
-   *   if there is no way for the caller to provide a reasonable destructor.
-   *   The destructors deletes the argument self as well.
+  /*!
+   * \brief Destructor - this should be called
+   * to destruct the manager_ctx  which backs the DLManagedTensor. It can be
+   * NULL if there is no way for the caller to provide a reasonable destructor.
+   * The destructors deletes the argument self as well.
    */
   void (*deleter)(struct DLManagedTensor * self);
 } DLManagedTensor;
+
+/*!
+ * \brief The device type in DLDevice.
+ */
+#ifdef __cplusplus
+typedef enum : uint64_t {
+#else
+typedef enum {
+#endif
+  /*! \brief bit mask to indicate that the tensor is read only. */
+  kDLFlagBitMaskReadOnly = 1UL,
+} DLFlagBitMask;
+
+/*!
+ * \brief A versioned and managed C Tensor object, manage memory of DLTensor.
+ *  This data structure is intended to facilitate the borrowing of DLTensor by
+ * another framework. It is not meant to transfer the tensor. When the borrowing
+ * framework doesn't need the tensor, it should call the deleter to notify the
+ * host that the resource is no longer needed.
+ *
+ * \note This is the current standard DLPack exchange data structure.
+ */
+struct DLManagedTensorVersioned {
+  /*!
+   * \brief The API and ABI version of the current managed Tensor
+   */
+  DLPackVersion version;
+  /*!
+   * \brief the context of the original host framework of
+   * DLManagedTensorVersined in which DLManagedTensorVersioned is used in the
+   * framework. It can also be NULL.
+   */
+  void *manager_ctx;
+  /*!
+   * \brief Destructor  - this should be called
+   * to destruct manager_ctx which holds the DLManagedTensorVersioned. It can
+   * be NULL if there is no way for the caller to provide a reasonable
+   * destructor. The destructors deletes the argument self as well.
+   */
+  void (*deleter)(struct DLManagedTensorVersioned *self);
+  /*!
+   * \brief Additional bitmask flags information about the tensor.
+   *
+   * By default the flags should be set to 0.
+   *
+   * \sa DLPackFlagBitMask
+   */
+  uint64_t flags;
```

Comment:
I agree

Reply:
Yes, this is the key part, and it would be useful to summarize the high-level idea in this file since this is the most important part of the spec. 

If we implement DLPack version `X` and get a tensor of version `Y` (where `Y>X`), what are we allowed to do? What are we not allowed to do?

Reply:
There are also two version numbers: the DLpack version, and the ABI version. Is the first one completely irrelevant, or does it also play a potential role? If yes, how is it different from the ABI version? If no, why does it need to be part of the ABI via a field stored in data structures?

Reply:
It could be a `major.minor` and minor allows e.g. additional (compatible!) flags or additional devices/dtypes (not strictly compatible, but clearly can't fail).  Not sure what the current scheme is?

Reply:
Added some clarification of the current scheme